lib/src/package_name.dart

```diff
@@ -153,39 +153,62 @@ class PackageRange extends PackageName {
   /// The allowed package versions.
   final VersionConstraint constraint;
 
-  /// The features that are required.
-  final Set<String> features;
+  /// Which features are required and which are opted out.
+  final Map<String, bool> features;
 
   /// Creates a reference to package with the given [name], [source],
   /// [constraint], and [description].
   ///
   /// Since an ID's description is an implementation detail of its source, this
   /// should generally not be called outside of [Source] subclasses.
   PackageRange(String name, Source source, this.constraint, description,
-      {Iterable<String> features})
+      {Map<String, bool> features})
       : features = features == null
-            ? const UnmodifiableSetView.empty()
-            : new UnmodifiableSetView(features.toSet()),
+            ? const {}
+            : new UnmodifiableMapView(new Map.from(features)),
         super._(name, source, description);
 
   PackageRange.magic(String name)
       : constraint = Version.none,
-        features = const UnmodifiableSetView.empty(),
+        features = const {},
         super._magic(name);
 
+  /// Returns a description of [features], or the empty string if [features] is
+  /// empty.
+  String get featureDescription {
+    if (features.isEmpty) return "";
+
+    var enabledFeatures = <String>[];
+    var disabledFeatures = <String>[];
+    features.forEach((name, enabled) {
+      (enabled ? enabledFeatures : disabledFeatures).add(name);
+    });
+
+    var description = "";
+    if (enabledFeatures.isNotEmpty) {
+      description += "with ${toSentence(enabledFeatures)}";
+      if (disabledFeatures.isNotEmpty) description += ", ";
+    }
+
+    if (disabledFeatures.isNotEmpty) {
+      description += "without ${toSentence(disabledFeatures)}";
+    }
+    return description;
+  }
+
   String toString() {
     if (isRoot) return "$name $constraint (root)";
```

Comment:
Can the root package enable/disable features? If so, probably don't want to early return here.